packages/server-wallet/src/protocols/ledger-manager.ts

```diff
@@ -55,108 +46,217 @@ export class LedgerManager {
-              await this.store.markLedgerRequests(defundedChannels, 'defund', 'succeeded', tx);
-
-              requiresAnotherCrankUponCompletion = true;
-              return;
-            }
-            default:
-              throw 'Unimplemented';
-          }
+    await this.store.lockApp(ledgerChannelId, async (tx, channel) => {
+      // TODO: Move these checks inside the DB query when fetching ledgers to process
+      if (!channel.protocolState.supported || channel.protocolState.supported.turnNum < 3) {
+        return;
+      }
+
+      let protocolState = await getProcessLedgerQueueProtocolState(this.store, ledgerChannelId, tx);
+
+      if (!hasUnhandledLedgerRequests(protocolState)) {
+        if (!requiresAnotherCrankUponCompletion) {
+          // pessimistically add state and proposal to outbox
+          await this.addStateAndProposalToOutbox(protocolState, response);
         }
-      });
-    }
+        response.queueChannelState(protocolState.fundingChannel);
+        return;
+      }
+
+      const crankAgain = await this.markIncludedRequestsAsComplete(protocolState, tx);
+      requiresAnotherCrankUponCompletion = requiresAnotherCrankUponCompletion || crankAgain;
+
+      // refresh the protocol state to be safe
+      // TODO: remove if possible
+      protocolState = await getProcessLedgerQueueProtocolState(this.store, ledgerChannelId, tx);
+
+      if (finishedExchangingProposals(protocolState)) {
+        const crankAgain2 = await this.exchangeSignedLedgerStates(protocolState, response, tx);
+        requiresAnotherCrankUponCompletion = requiresAnotherCrankUponCompletion || crankAgain2;
+      }
+
+      // refresh the protocol state to be safe
+      // TODO: remove if possible
+      protocolState = await getProcessLedgerQueueProtocolState(this.store, ledgerChannelId, tx);
+
+      // exchange proposals - when do I need to do this?
+      const crankAgain3 = await this.exchangeProposals(protocolState, tx);
+      requiresAnotherCrankUponCompletion = requiresAnotherCrankUponCompletion || crankAgain3;
+
+      // refresh the protocol state to be safe
+      // TODO: remove if possible
+      protocolState = await getProcessLedgerQueueProtocolState(this.store, ledgerChannelId, tx);
+
+      if (!requiresAnotherCrankUponCompletion) {
+        // pessimistically add state and proposal to outbox
+        await this.addStateAndProposalToOutbox(protocolState, response);
+      }
+      response.queueChannelState(protocolState.fundingChannel);
+    });
+
     return requiresAnotherCrankUponCompletion;
   }
+
+  private async signLedgerUpdate(
+    fundingChannel: ChannelStateWithSupported,
+    stateToSign: StateVariables,
+    response: WalletResponse,
+    tx: Transaction
+  ) {
+    const {myIndex, channelId} = fundingChannel;
+    const channel = await Channel.forId(channelId, tx);
+    const signedState = await this.store.signState(channel, stateToSign, tx);
```

Comment:
Why are we fetching the channel from the DB again if we have a `fundingChannel` available?

Reply:
This was just a direct translation of what was there before:
```ts
  case 'SignLedgerUpdate': {	
      const {myIndex, channelId} = protocolState.fundingChannel;	
      const channel = await Channel.forId(channelId, tx);	
      const signedState = await this.store.signState(channel, action.stateToSign, tx);	
      response.queueState(signedState, myIndex, channelId);	
      return;	
    }
```